### ISSUE TITLE
Clarify the complexity of `dequeue()`

### ISSUE THREAD

**Issue by alexypdu** [datastructures-js/queue#26]
Hello, I wonder why dequeuing all elements runs in $O(n \log n)$? The following operation will only run $\log n$ times:

```js
// only remove dequeued elements when reaching half size
// to decrease latency of shifting elements.
this._elements = this._elements.slice(this._offset);
```

And since `Array.slice()` runs in $O(n)$, the complexity should be: $$\frac{n}{2} + \frac{n}{4}  + \cdots + 1 < \frac{n}{2} + \frac{n}{4}  + \cdots = \frac{n}{2} \cdot \sum_{i=0}^{\infty} \frac{1}{2^i} = \frac{n}{2} \cdot 2 = n = O(n)$$

So the overall complexity of dequeuing all elements should be $O(n + n) = O(n)$, and hence the amortized complexity of `dequeue()` is $O(1)$. Please correct me if I am wrong :)

**Comment by eyas-ranjous** [comment]
that's a good analysis @alexypdu indeed that's an O(1) operation. I haven't spent much time in analyzing the performance when I wrote the Readme. Feel free to make the change, and add the explanation to it as well and assign the pr to me.

**Comment by eyas-ranjous** [comment]
@alexypdu can you check the pr https://github.com/datastructures-js/queue/pull/27 and let me know if it looks good?

**Comment by alexypdu** [comment]
I would probably rephrase it to something like this:

Internally, when half the elements have been dequeued, we will resize the dynamic array using `Array.slice()` which runs in $O(n)$. Since dequeuing all $n$ elements will resize the array $\log_2n$ times, the complexity is $$1 + 2 + 4 + \cdots + 2^{\log_2 n - 1} = 2 ^ {(\log_2 n - 1) + 1} - 1 = n - 1 = O(n)$$ Hence the overall complexity of dequeuing all elements is $O(n + n) = O(n)$, and the amortized complexity of `dequeue()` is thus $O(1)$.
